src/java.net.http/share/classes/jdk/internal/net/http/WindowUpdateSender.java

```diff
@@ -117,18 +117,24 @@ abstract class WindowUpdateSender {
      *             the caller wants to buffer.
      */
     boolean canBufferUnprocessedBytes(int len) {
-        return !checkWindowSizeExceeded(unprocessed.addAndGet(len));
+        long buffered, rcv;
+        // get received before unprocessed in order to avoid counting
+        // unprocessed bytes that might get unbuffered asynchronously
+        // twice.
+        rcv = received.get();
+        buffered = unprocessed.addAndGet(len);
+        return !checkWindowSizeExceeded(rcv, buffered);
     }
 
     // adds the provided amount to the amount of already
     // received and processed bytes and checks whether the
     // flow control window is exceeded. If so, take
     // corrective actions and return true.
-    private boolean checkWindowSizeExceeded(long len) {
+    private boolean checkWindowSizeExceeded(long received, long len) {
         // because windowSize is bound by Integer.MAX_VALUE
         // we will never reach the point where received.get() + len
```

Comment:
Nit - the method param `received` shadows the class level field and the comment that says "received.get()" would need to be reworded too. It might be better to rename the method param to something else.

Reply:
I guess I could rename it to `processed`.